lnd: use persisted node announcement data from previous runtime

Abdulkbk opened the pull request:
closes #7123 

replaces #8690 

## Change Description
Description of change / link to associated issue.
In this commit, the node announcement config is retrieved from the database and reused when a node restarts.

We will begin by checking the configuration set in lnd.conf  file or passed as arguments when the node starts. If the configuration is not present, we will check for the ones saved on disk and use them. If none are found, we will resort to using the default configurations.

## Steps to Test
Steps for reviewers to follow to test the change.
- Make sure no alias or color is set in your lnd.conf (config here takes precedence)
- Use the `updatenodeannouncement` RPC call to update your node alias, color, addresses or feature bits. 
- Verify the alias is set using lncli getinfo
- Now stop and start the node again and run lncli getinfo again
You'll notice that the fields you update earlier are persisted


## Pull Request Checklist
### Testing

- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

ellemouton commented:
@Abdulkbk  - thanks! Although I think you could have just updated the commits in the original PR - no need for a new PR :)

Abdulkbk commented:
> @Abdulkbk  - thanks! Although I think you could have just updated the commits in the original PR - no need for a new PR :)

At first, I thought about doing that, but then I ended up creating a new PR. Thanks for the feedback; I'm really learning a lot.

Abdulkbk commented:
Hi @alexbosworth, I've opened this PR to fix the issue #7123 you reported. Could you spare some time to review it? Your input will help me improve the solution.

alexbosworth reviewed (approved):
utconcept ack

Chinwendu20 commented on server.go:
```diff
@@ -829,7 +829,7 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 		// If the color is still set to the default, we'll use the
 		// persisted color.
-		if cfg.Color == "#3399FF" {
+		if cfg.Color == defaultColor {
```
I think you can create a separate commit where you export `defaultColor`. Then use it in your previous commit.  Then maybe delete this commit. `git rebase` can help.

Abdulkbk replied:
Yes, I agree. I'll squash it into one commit.

Chinwendu20 commented on server.go:
```diff
@@ -803,33 +804,76 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
+
+		// If an alias was not specified in the configuration,
+		// we'll use the saved alias.
+		if alias == "" {
```
I think alias would always be an empty string so this not needed.

Chinwendu20 commented on server.go:
```diff
@@ -803,33 +804,76 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
+
+		// If an alias was not specified, we'll generate one based on
+		// the serialized public key.
+		if alias == "" {
```
I do not think there alias  never not be an empty string here?

Chinwendu20 commented on server.go:
```diff
@@ -803,33 +804,76 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 
+	// Fetch the source node from the channel graph. If a source node is
+	// found, we'll use it's values for our initial node announcement.
+	persistedNode, err := chanGraph.SourceNode()
```
cool implementation using the lightning node bucket 🥳🥳🥳

Chinwendu20 commented on server.go:
```diff
@@ -803,33 +804,76 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
+		}
+
+		features = s.featureMgr.Get(feature.SetNodeAnn)
+
```
Remove extra space

Chinwendu20 reviewed (commented):
Nice work. Left some suggestions. I think the commit message does not follow guideline (i.e. length of the message):
https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#model-git-commit-messages

Here 6ebec6148ef072bd5cddc7, I do not think this is necessary:
'''
lnd: refactor code to elliminate duplicate nill check

lnd: fix comment
'''

lightninglabs-deploy commented:
@abdulkbk, remember to re-request review from reviewers when ready

Abdulkbk commented:
> Also I think this needs a release note? Maybe add here: https://github.com/lightningnetwork/lnd/blob/7065b6462edf48f3c86a58f2fd215cb0db2c0474/docs/release-notes/release-notes-0.18.1.md

Since this PR persists whatever was previously updated with the `updatenodeannouncement` lncli command, it means I should add a bullet point to explain that, right? @Chinwendu20

Chinwendu20 commented:
I think you can add a bullet point to talk about what the change actually is and maybe a short explanation. Maybe under [functional updates](https://github.com/lightningnetwork/lnd/blob/7065b6462edf48f3c86a58f2fd215cb0db2c0474/docs/release-notes/release-notes-0.18.1.md#functional-updates)

Abdulkbk commented:
> I think you can add a bullet point to talk about what the change actually is and maybe a short explanation. Maybe under [functional updates](https://github.com/lightningnetwork/lnd/blob/7065b6462edf48f3c86a58f2fd215cb0db2c0474/docs/release-notes/release-notes-0.18.1.md#functional-updates)

I just noticed that the [issue](https://github.com/lightningnetwork/lnd/issues/7123) this PR is addressing is under milestone [0.19.0](https://github.com/lightningnetwork/lnd/milestone/51). Shouldn't the milestone version and release notes version be the same?

Chinwendu20 commented:
I think it makes sense that it should be the same.

lightninglabs-deploy commented:
@abdulkbk, remember to re-request review from reviewers when ready

alexbosworth commented on server.go:
```diff
@@ -803,33 +804,75 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
 	chanGraph := dbs.GraphDB.ChannelGraph()
 
+	// Fetch the source node from the channel graph. If a source node is
+	// found, we'll use it's values for our initial node announcement.
```
```suggestion
	// found, we'll use its values for our initial node announcement.
```

Abdulkbk replied:
Thank you for bringing that to my attention.

Abdulkbk commented:
Hi @yyforyongyu , I noticed that you are one of the code owners of [netann](https://github.com/lightningnetwork/lnd/blob/f464dac1cd2f63af4b4e3de4a2c23354acf8db6e/.github/CODEOWNERS-HINT#L144). This PR touches `nodeannouncement`, which I think falls under that. Would you mind reviewing it? It is required before I can get the tests to run.

yyforyongyu commented on server.go:
```diff
@@ -803,33 +804,75 @@ func newServer(cfg *Config, listenAddrs []net.Addr,
+	)
+
+	// If the node was previously persisted, we'll use the saved values.
+	if persistedNode != nil {
```
haven't checked the details, but I think we should create a method instead of growing the already giant `newServer` method.

Abdulkbk replied:
I agree. I've refactored the code into a method

yyforyongyu reviewed (commented):
Approved CI run.

lightninglabs-deploy commented:
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

Abdulkbk commented:
Could you help approve the CI, @yyforyongyu?

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

yyforyongyu commented:
Restarting a node in the itest can be done in multiple ways,

for instance you can suspend a node and restart it later,
https://github.com/lightningnetwork/lnd/blob/a388c1f39d849005b58eae81516b4104929101a6/itest/lnd_channel_backup_test.go#L797

or just restart it,
https://github.com/lightningnetwork/lnd/blob/a388c1f39d849005b58eae81516b4104929101a6/itest/lnd_misc_test.go#L209

Abdulkbk commented:
> Restarting a node in the itest can be done in multiple ways,
> 
> 
> 
> for instance you can suspend a node and restart it later,
> 
> https://github.com/lightningnetwork/lnd/blob/a388c1f39d849005b58eae81516b4104929101a6/itest/lnd_channel_backup_test.go#L797
> 
> 
> 
> or just restart it,
> 
> https://github.com/lightningnetwork/lnd/blob/a388c1f39d849005b58eae81516b4104929101a6/itest/lnd_misc_test.go#L209

Thanks for the pointers 👍

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@chinwendu20: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

yyforyongyu commented:
!lightninglabs-deploy mute

Abdulkbk commented:
Hmmm can't seem to figure out why the entire itests keep [failing ](https://github.com/lightningnetwork/lnd/actions/runs/15401312314/job/43335014417?pr=8825), will need to dig in a little deeper...

Abdulkbk commented:
> With the following error in Carol's log:
> 
> ```
> unable to create server: can't unset feature bit 25 (route-blinding): feature is used in standard protocol set
> ```

Thanks for pointing that out. This made me realize I could download and view a particular node's logs from the CI, very helpful.


I was initially updating the feature manager with the feature sets that were persisted in the previous run, something like:
```go
        prevFeatures := map[feature.Set]*lnwire.RawFeatureVector{
		feature.SetNodeAnn: sourceNode.Features.RawFeatureVector,
	}
	err = s.featureMgr.UpdateFeatureSets(prevFeatures)
	if err != nil {
		srvrLog.Errorf("unable to update feature sets: %v", err)

		return err
	}
```
This ensures that when you add an unknown feature bit (2, for example) before restarting, you get to see it in `lncli getinfo` response even after a restart.
```bash
...
 "2":  {
            "name":  "unknown",
            "is_required":  true,
            "is_known":  false
        },
...
```

The issue arises when we start a node carol in this case with `--protocol.no-route-blinding` and later restart it with route blinding support, as is done in the itest [here](https://github.com/lightningnetwork/lnd/blob/master/itest/lnd_route_blinding_test.go#L1034), the initial implementation will try to unset the feature bit (since it was not used in the last run).

Abdulkbk commented:
@saubyk could you add this to the review board?

ellemouton commented on itest/lnd_channel_graph_test.go:
```diff
@@ -654,6 +655,98 @@ func testUpdateNodeAnnouncement(ht *lntest.HarnessTest) {
+		Alias: "alice",
+		Color: "#eeeeee",
+		AddressUpdates: []*peersrpc.UpdateAddressAction{
+			{
```
it would be good to test that we can still remove an address too

Abdulkbk replied:
Now testing for this behaviour

ellemouton commented on server.go:
```diff
@@ -5573,3 +5584,60 @@ func (s *server) AttemptRBFCloseUpdate(ctx context.Context,
+
+// getNodeAnnouncementFields populates the alias, color, and addresses. This
+// is used during startup to construct the initial node announcement.
+func getNodeAnnouncementFields(ctx context.Context, cfg *Config, s *server,
+	serializedPubKey [33]byte) (color.RGBA, string, []net.Addr, error) {
```
rather make a method on `server`

ie: 
```
func (s *server) getNodeAnnFields(ctx context.Context, serializedPubKey route.Vertex) ...
```

Abdulkbk replied:
Nice, looks cleaner!

ellemouton commented on server.go:
```diff
@@ -5573,3 +5584,60 @@ func (s *server) AttemptRBFCloseUpdate(ctx context.Context,
+	// config color is not changed from default (#3399FF).
+	color, err = lncfg.ParseHexColor(cfg.Color)
+	if err != nil {
+		srvrLog.Errorf("unable to parse color: %v\n", err)
+
+		return color, "", nil, err
```
rather just return a decorated error - it will get logged higher up

ellemouton commented on server.go:
```diff
@@ -932,20 +933,30 @@ func newServer(ctx context.Context, cfg *Config, listenAddrs []net.Addr,
-	alias := cfg.Alias
-	if alias == "" {
-		alias = hex.EncodeToString(serializedPubKey[:10])
+	// Create a map to track existing addresses for quick lookup. This is
```
can probs move this logic into the getNodeAnnFields method and just pass it `selfNodes` and then used the returned `addresses` as the new `selfNodes`

Abdulkbk replied:
All logic has been moved.

ellemouton reviewed (commented):
Nice work! Will be great to get this in as it is closely tied to https://github.com/lightningnetwork/lnd/issues/9999 (where we also will need to start fetching our SourceNode from the DB on startup

Abdulkbk commented on server.go:
```diff
@@ -5573,3 +5584,60 @@ func (s *server) AttemptRBFCloseUpdate(ctx context.Context,
+	// config values.
+	sourceNode, err := s.graphDB.SourceNode(ctx)
+	if err != nil {
+		srvrLog.Debugf("Unable to get source node from graphDB to "+
+			"construct initial node announcement: %v", err)
+
```
Done

Abdulkbk commented:
Thanks for the review @ellemouton I've addressed the feedback you left.

Abdulkbk commented:
I noticed some of the itests are failing particulary with this error from [here](https://github.com/lightningnetwork/lnd/blob/master/graph/db/sql_store.go#L468)
```bash
[ERR] LTND lnd.go:169: Shutting down due to error in main method err="unable to fetch source node: unable to fetch V1 source node: source node does not exist"
```
will investigate more...

ellemouton commented:
@Abdulkbk  - that is fixed in https://github.com/lightningnetwork/lnd/issues/9999

I think it's better to build this PR on top of that one since that one will probably be merged first

Abdulkbk commented:
> @Abdulkbk - that is fixed in #9999
> 
> I think it's better to build this PR on top of that one since that one will probably be merged first

Alright, I will rebase on top of that once it lands.

Abdulkbk commented on server.go:
```diff
@@ -929,28 +930,6 @@ func newServer(ctx context.Context, cfg *Config, listenAddrs []net.Addr,
-		return nil, err
-	}
-
 	// TODO(elle): All previously persisted node announcement fields (ie,
```
Cool, removed the TODO and moved the handling to the helper function.

ellemouton commented:
\gemini review

Abdulkbk commented on server.go:
```diff
@@ -976,10 +955,25 @@ func newServer(ctx context.Context, cfg *Config, listenAddrs []net.Addr,
 		LastUpdate:           nodeLastUpdate,
-		Addresses:            selfAddrs,
+		Addresses:            addresses,
 		Alias:                nodeAlias.String(),
```
Indeed, done.

ellemouton commented on itest/lnd_channel_graph_test.go:
```diff
@@ -654,6 +655,120 @@ func testUpdateNodeAnnouncement(ht *lntest.HarnessTest) {
+		"color":     1,
+		"addresses": 2,
+	}
+	assertUpdateNodeAnnouncementResponse(ht, response, expectedOps)
```
oh cool, i see you do this via `alice.RPC.GetInfo()` below :+1: perhaps makes sense to have that as a helper earlier on and call it here too

ellemouton commented on itest/lnd_channel_graph_test.go:
```diff
@@ -654,6 +655,120 @@ func testUpdateNodeAnnouncement(ht *lntest.HarnessTest) {
+// testNodeAnnouncementPersistence tests that the node announcement configs are
+// persisted correctly and reused when the node is restarted using the correct
+// hierarchy (config > source node > defaults).
+func testNodeAnnouncementPersistence(ht *lntest.HarnessTest) {
```
very nice test!!

ellemouton commented on server.go:
```diff
@@ -931,95 +932,20 @@ func newServer(ctx context.Context, cfg *Config, listenAddrs []net.Addr,
+	// We now get the self node which represents our node on the channel
+	// graph.
+	selfNode, err := s.getSelfNode(
+		ctx, selfAddrs, nodeKeyDesc, serializedPubKey,
```
actually, you can probably just pass `listenAddrs` in directly and do all the addr parsing within the helper.

ellemouton replied:
also, note that you are passing in both `nodeKeyDesc` and `serializedPubKey` but `serializedPubKey` is computed from `nodeKeyDesc` - so both should not be needed

Abdulkbk replied:
> actually, you can probably just pass `listenAddrs` in directly and do all the addr parsing within the helper.

Thought about passing `listenAddrs`, but I noticed we use it in the `natTraversal` check to extract the port that we use to configure `portforwarding`, which then returns a list of IPs that we append to `externalIPStrings`. So I'm passing in externalIPStrings instead in bcea84debfd8ca0ad59f3abff26c47c0eb286cff. Let me know if that makes sense.

ellemouton commented on server.go:
```diff
@@ -931,95 +932,20 @@ func newServer(ctx context.Context, cfg *Config, listenAddrs []net.Addr,
-		selfNode.AuthSigBytes,
+	// We now get the self node which represents our node on the channel
+	// graph.
+	selfNode, err := s.getSelfNode(
```
`get*` implies that this purely does a read. I suggest rather naming it something like `generateSelfNode`/`buildSelfNode`/`setSelfNode` to indicate that it is doing some processing given its arguments and also some persistence.

ellemouton replied:
note that the only thing used from the returned `selfNode` is the public key bytes which we actually know before ever calling this. So i dont think the new helper needs to return anything

ellemouton commented on server.go:
```diff
@@ -5614,3 +5540,132 @@ func (s *server) AttemptRBFCloseUpdate(ctx context.Context,
+	// With the announcement generated, we'll sign it to properly
+	// authenticate the message on the network.
+	authSig, err := netann.SignAnnouncement(
+		s.nodeSigner, nodeKeyDesc.KeyLocator, nodeAnn,
```
can use `s.identityKeyLoc`

ellemouton commented on server.go:
```diff
@@ -5614,3 +5540,132 @@ func (s *server) AttemptRBFCloseUpdate(ctx context.Context,
 	return updates, nil
 }
+
+// getSelfNode returns the self node. It handles setting the node announcement,
+// signing it, and updating the source node in the graph.
```
the comment should speak to the fact that config values that differ from defaults are prio'd, otherwise pre-existing settings found on disk are prio'd

ellemouton commented on itest/list_on_test.go:
```diff
@@ -166,6 +166,10 @@ var allTestCases = []*lntest.TestCase{
 		TestFunc: testUpdateNodeAnnouncement,
 	},
+	{
+		Name:     "node announcement persistence",
+		TestFunc: testNodeAnnouncementPersistence,
+	},
```
suggest: `self node announcement persistence` (same for test name)

Abdulkbk commented on itest/lnd_channel_graph_test.go:
```diff
@@ -654,6 +655,120 @@ func testUpdateNodeAnnouncement(ht *lntest.HarnessTest) {
+		for _, addr := range targetAddrs {
+			_, ok := addrs[addr]
+			if !ok {
+				return fmt.Errorf("address %v not found in "+
+					"node announcement", addr)
```
Are you sure @gemini-code-asist? I count 77.

Abdulkbk commented:
> Here is a suggested patch - let me know what you think :)
> 
> [node_ann.patch](https://github.com/user-attachments/files/21252237/node_ann.patch)

Thanks for the review @ellemouton. I am unable to access the patch, but I've addressed your feedback and left a [question](https://github.com/lightningnetwork/lnd/pull/8825#discussion_r2219776519).

lightninglabs-deploy commented:
@yyforyongyu: review reminder
@abdulkbk, remember to re-request review from reviewers when ready

ellemouton commented on server.go:
```diff
@@ -5631,3 +5549,136 @@ func (s *server) AttemptRBFCloseUpdate(ctx context.Context,
+	serializedPubKey [33]byte) error {
+
+	var (
+		color          color.RGBA
```
nit: you dont need to define color here.

ellemouton commented on server.go:
```diff
@@ -919,21 +919,8 @@ func newServer(ctx context.Context, cfg *Config, listenAddrs []net.Addr,
 		}
```
note: this is still in a fixup commit

ellemouton reviewed (approved):
nice work!
